os/arch/arm/src/amebad/amebad_timer_lowerhalf.c

```diff
@@ -0,0 +1,485 @@
+static int amebad_gpt_getstatus(struct timer_lowerhalf_s *lower, struct timer_status_s *status);
+static int amebad_gpt_settimeout(struct timer_lowerhalf_s *lower, uint32_t timeout);
+static void amebad_gpt_setcallback(struct timer_lowerhalf_s *lower, CODE tccb_t callback, void *arg);
+static int amebad_gpt_ioctl(struct timer_lowerhalf_s *lower, int cmd, unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+static const struct timer_ops_s g_timer_ops = {
+	.start = amebad_gpt_start,
+	.stop = amebad_gpt_stop,
+	.getstatus = amebad_gpt_getstatus,
+	.settimeout = amebad_gpt_settimeout,
+	.setcallback = amebad_gpt_setcallback,
+	.ioctl = amebad_gpt_ioctl,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt0_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt1_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt2_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt3_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name: amebad_gpt_handler
+ *
+ * Description:
+ *   timer interrupt handler
+ *
+ * Input Parameters:
+ *
+ * Returned Values:
+ *
+ ****************************************************************************/
+static void amebad_gpt_handler(uint32_t data)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)data;
+	DEBUGASSERT(priv);
+	uint32_t next_interval_us = 0;
+
+	if (priv->freerunmode) {	/* Free Run Mode */
+		priv->freeruncount++;
+	} else {					/* Alarm Mode */
+		if (priv->callback(&next_interval_us, priv->arg)) {
+			if (next_interval_us > 0) {
+				priv->gpt_timeout = next_interval_us;
+				gtimer_reload(&priv->obj, priv->gpt_timeout);
+			}
+		} else {
+			amebad_gpt_stop((struct timer_lowerhalf_s *)priv);
+		}
+	}
+}
+
+/****************************************************************************
+ * Name: amebad_gpt_start
+ *
+ * Description:
+ *   Start the timer, resetting the time to the current timeout,
+ *
+ * Input Parameters:
+ *   lower - A pointer the publicly visible representation of the "lower-half"
+ *           driver state structure.
+ *
+ * Returned Values:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+static int amebad_gpt_start(struct timer_lowerhalf_s *lower)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)lower;
+	DEBUGASSERT(priv);
+
+	if (!priv->started) {
+		if (priv->freerunmode) {	/* Free Run Mode */
+			priv->obj.handler = (void *)amebad_gpt_handler;
```

Comment:
You set this twice, one is here, another is in amebad_gpt_setcallback. I think you could remove duplicated.

Reply:
I have checked it again, this duplicate is necessary. It happens that callback function is called before timer is started. So that 'callback is not NULL but priv->started is false'(Line343). In this case, the callback is assigned but the handler is not assigned. If the timer starts then, the handler need to be assigned again(Line186).